### PR TITLE
Reduce overall memory usage by removing cgi dependency and writing a protected function

### PR DESCRIPTION
**Pull request by DemetriosBairaktaris**
In response to issue #4313 , I removed the cgi module import and added a parse_header function nested inside get_encoding_from_headers in utils.py.  I have run the tests both before and after making these changes.  The test_util cases all passed before and they still do now.  Please let me know about any suggestions to this code.

### CODE REVIEW

File: requests/utils.py [hunk 1]
```diff
@@ -453,13 +452,27 @@ def get_encoding_from_headers(headers):
...
+        content_type_and_params_delimiter = ';'
+        content_type += content_type_and_params_delimiter
+
+        tokens = content_type.split(content_type_and_params_delimiter)
+        content_type_index = 0
+        params_index = 1
+
+        content_type = tokens[content_type_index]
+        params = tokens[params_index]
+        params_dict = dict(param.split('=') for param in params.split())
```
Comment (sigmavirus24) [inline requests/utils.py]: I'm not convinced this is going to be correct. I think having an `=` inside the parameter value is acceptable if it's quoted. We should probably then use `param.split('=', 1)` to ensure we only split on the first found =.

File: requests/utils.py [hunk 2]
```diff
@@ -453,13 +452,27 @@ def get_encoding_from_headers(headers):
...
+        return content_type,params_dict
```
Comment (sigmavirus24) [inline requests/utils.py]: Please add a space, e.g.,

```py
return content_type, params_dict
```

File: requests/utils.py [hunk 3]
```diff
@@ -453,13 +452,27 @@ def get_encoding_from_headers(headers):
     :param headers: dictionary to extract encoding from.
     :rtype: str
     """
+    def parse_header(content_type):
+        #Inner function to parse header
```
Comment (sigmavirus24) [inline requests/utils.py]: Why does this need to be an inner function? Why can't we define it at the module level as `_parse_header`?

File: requests/utils.py (hunk 2 above)
Comment (DemetriosBairaktaris) [reply requests/utils.py]: Thanks for your suggestions.

File: requests/utils.py (hunk 1 above)
Comment (DemetriosBairaktaris) [reply requests/utils.py]: Ill check it out. Thank you

**Comment by DemetriosBairaktaris** [comment]
@sigmavirus24 Thoughts on this?  I figured since the since we are only worrying about parsing content-type headers, we should focus the function on that format.

File: requests/utils.py [hunk 4]
```diff
@@ -459,7 +488,7 @@ def get_encoding_from_headers(headers):
     if not content_type:
         return None
 
-    content_type, params = cgi.parse_header(content_type)
+    content_type, params = _parse_content_type_header(content_type)
```
Comment (sigmavirus24) [inline requests/utils.py:486]: If `content_type` is checked above and this is a private function, do we need the guard at the top of the function above? If the check above is ever removed, we could end up with a `TypeError` when we  effectively do `content_type, params = None`.

File: requests/utils.py [hunk 5]
```diff
@@ -447,6 +446,36 @@ def get_encodings_from_content(content):
...
+    """Returns content type and parameters from given header
+
+    :param header: string
+    :return: tuple containing content type and dictionary of
+         parameters
+    """
+    if not header:
+        return None
+    # append delimiter on end to ensure at least two elements when split by ';'
+    header += ';'
```
Comment (sigmavirus24) [inline requests/utils.py]: I don't think this is necessary. Can we do

```py
tokens = header.split(';', 1)
content_type, params_list = tokens[0].strip(), tokens[1:]
if  len(params_list) < 1:
    return content_type, {}
params = params_list[0]
```

If I understand correctly that's the effect of this otherwise. It's simple, clearer, and doesn't require us to string concatenation which results in a new allocation and isn't terribly efficient.

```py
>>> a = [1]
>>> b, c = a[0], a[1:]
>>> b, c
(1, [])
```

File: requests/utils.py [hunk 6]
```diff
@@ -447,6 +446,36 @@ def get_encodings_from_content(content):
...
+    # split content type's main value from params
+    tokens = header.split(';', 1)
+    content_type_index = 0
+    params_index = 1
+
+    content_type = tokens[content_type_index].strip()
+    params = tokens[params_index]
+    params_dict = dict()
+
+    for param in params.split(';'):
```
Comment (sigmavirus24) [inline requests/utils.py]: The above said, if we weren't restricting it to split once we could simplify the code further and do:

```
tokens = header.split(';')
content_type, params = tokens[0].strip(), tokens[1:]
params_dict = {}. # Using dict is actually slower than a dictionary literal. Weird but tru
for param in params:
    # ... What you have below

return content_type, params_dict
```

File: requests/utils.py (hunk 5 above)
Comment (DemetriosBairaktaris) [reply requests/utils.py]: That is pretty nice.  Didn't know a[1:] would not throw an index out of bounds... nice.

**Comment by DemetriosBairaktaris** [comment]
@sigmavirus24 Ok, I pushed some changes with your suggestions, it's looking better for sure.

File: requests/utils.py [hunk 7]
```diff
@@ -447,6 +446,29 @@ def get_encodings_from_content(content):
...
+    :return: tuple containing content type and dictionary of
+         parameters
+    """
+
+    tokens = header.split(';')
+    content_type, params = tokens[0].strip(), tokens[1:]
+    params_dict = {} 
+
+    for param in params:
+        if param and not param.isspace():
```
Comment (sethmlarson) [inline requests/utils.py]: Can we not call `param = param.strip()` before the `if param` and save a call to `param.isspace()`?

ie Is this:
```python
for param in params:
    param = param.strip()
    if param:
        # Continue here...
```
the same as this
```python
for param in params:
    if param and not param.isspace():
        param = param.strip()
        # Continue here...
```
in all situations?

File: requests/utils.py [hunk 8]
```diff
@@ -447,6 +446,29 @@ def get_encodings_from_content(content):
...
+            param = param.strip()
+            key, value = param, True
+            if '=' in param:
```
Comment (sethmlarson) [inline requests/utils.py]: Might be able to get rid of the list comprehension here by using `param.find('=')` and checking for a non-negative? Definitely would be faster. Something to the tune of:

```python
key, value = param, True
index = param.find('=')
if index != -1:
    key = param[:index].strip('\'" ')
    value = param[index+1:].strip('\'" ')
```

File: tests/test_utils.py [hunk 9]
```diff
@@ -470,6 +470,33 @@ def test_parse_dict_header(value, expected):
...
+        (
+            'application/json ; charset=utf-8',
+            ('application/json', {'charset': 'utf-8'})
+        ),
+        (
+            'text/plain',
+            ('text/plain', dict())
+        ),
+        (
+            'multipart/form-data; boundary = something ; \'boundary2=something_else\' ; no_equals ',
```
Comment (sethmlarson) [inline tests/test_utils.py]: Can you also add a test case with double-quotes around the value? ie: `boundary = "something";` should return `{'boundary': 'something'}`.

**Comment by DemetriosBairaktaris** [comment]
@SethMichaelLarson Good suggestions, I pushed some changes.

File: requests/utils.py [hunk 10]
```diff
@@ -456,15 +456,19 @@ def _parse_content_type_header(header):
...
+        param = param.strip()
+        if param:
             key, value = param, True
-            if '=' in param:
-                param_tokens = [x.strip('\'" ') for x in param.split('=', 1)]
-                key, value = param_tokens[0], param_tokens[1]
+            index_of_equals = param.find("=")
+            if index_of_equals != -1:
+                before_equals = slice(0, index_of_equals)
+                after_equals = slice(index_of_equals + 1, len(param))
```
Comment (SethMichaelLarson) [inline requests/utils.py]: Can you remove the `slice`s here? Just use index slice notation, (`param[:index_of_equals]` and `param[index_of_equals:]`) much more readable IMO.

File: requests/utils.py [hunk 11]
```diff
@@ -447,6 +446,33 @@ def get_encodings_from_content(content):
...
+    params_dict = {}
+    items_to_strip = "\"' "
+
+    for param in params:
+        param = param.strip()
+        if param:
+            key, value = param, True
+            index_of_equals = param.find("=")
+            if index_of_equals != -1:
+                before_equals = slice(0, index_of_equals)
```
Comment (sethmlarson) [inline requests/utils.py]: Can you remove the `slice()` usage here? Using bracket-slice notation is more readable and faster.

`param[:index_of_equals]` and `param[index_of_equals+1:]`

**Comment by sethmlarson** [comment]
Besides my two last comments this looks good to me. I'll leave this review for @sigmavirus24 now. :)

File: tests/test_utils.py (hunk 9 above)
Comment (sethmlarson) [reply tests/test_utils.py]: Still would like to see a test-case containing this scenario.

**Comment by sigmavirus24** [comment]
Thanks for your patience @DemetriosBairaktaris ! This looks excellent!
